hw/ip/otbn/util/shared/information_flow.py

```diff
@@ -0,0 +1,557 @@
+        self.flag = flag
+        self.constant_dependent = False
+
+    def evaluate(self, op_vals: Dict[str, int],
+                 constant_regs: Dict[str, int]) -> str:
+        if 'flag_group' not in op_vals:
+            raise ValueError(
+                'Operand flag_group not found in provided operand values: {}'.
+                format(op_vals.keys()))
+        return 'fg{}-{}'.format(op_vals['flag_group'], self.flag)
+
+
+class InsnConstantNode(InsnInformationFlowNode):
+    '''Represents instruction node whose value does not depend on operands.'''
+    def __init__(self, node: str):
+        self.node = node
+        self.constant_dependent = False
+
+    def evaluate(self, op_vals: Dict[str, int],
+                 constant_regs: Dict[str, int]) -> str:
+        return self.node
+
+
+def _parse_iflow_nodes(
+        node: str, what: str,
+        operands: List[Operand]) -> List[InsnInformationFlowNode]:
+    '''Parses information flow node(s) from the instruction description.
+
+    Valid information flow nodes are one of the following:
+    - the name of a *register* operand from the operands list
+    - "wref-<reg operand name>" for instructions where a WDR is indirectly
+      accessed via a GPR; in this case <reg operand name> is the GPR and must
+      be in the operands list
+    - one of the special strings "dmem", "acc", or "mod"
+    - a flag (represented as "<flag group>-<flag>", where <flag group> can
+      be either "fg0", "fg1", or (if the instruction has a flag_group
+      operand) simply "flags" to select the current flag group, and <flag>
+      can be l, c, m, z, or "all".
+
+    Raises a ValueError if the node does not have a valid format.
+    '''
+    node = node.lower()
+
+    # Check if node is a register operand
+    for op in operands:
+        if node == op.name and isinstance(op.op_type, RegOperandType):
+            return [InsnRegOperandNode(op)]
+
+    # Check if node is an indirect reference to a WDR through a GPR
+    if node.startswith('wref-'):
+        gpr = node[len('wref-'):]
+        for op in operands:
+            if (gpr == op.name and isinstance(op.op_type, RegOperandType) and
+                    op.op_type.reg_type == 'gpr'):
+                return [InsnIndirectWDRNode(op)]
+
+    # Check if node is a special string
+    if node == 'dmem' or node in SPECIAL_REG_NAMES:
+        return [InsnConstantNode(node)]
+
+    # Try to interpret node as a flag or set of flags
+    node_split = node.split('-')
+    if len(node_split) != 2:
+        print([op.name for op in operands])
+        raise ValueError('Cannot parse information flow node {} for {}'.format(
+            node, what))
+    flag_group_str, flag = node_split
+
+    # Case where flag group depends on the flag_group operand
+    if flag_group_str == 'flags':
+        if flag == 'all':
+            return [InsnGroupFlagNode(flag) for flag in FLAG_NAMES]
+        return [InsnGroupFlagNode(flag)]
```

Comment:
Do we need to check that `flag in FLAG_NAMES` here? Otherwise what happens if I type `flags-foo`?

Reply:
That gets checked in the `InsnGroupFlagNode` constructor, which raises an error if `flag` isn't in `FLAG_NAMES`.